src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -274,40 +293,100 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
             }
             constantSourcesBuilder.Free();
 
-            // Run a syntax walk if any of the generators requested it
-            if (receiverCount > 0)
+
+            // PROTOTYPE(source-generators): we don't need to run at all if none of the inputs have changes.
+            var driverStateBuilder = new DriverStateTable.Builder(_state.StateTable, cancellationToken);
+            driverStateBuilder.SetInputState(SharedInputNodes.Compilation, NodeStateTable<Compilation>.WithSingleItem(compilation, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.AnalzerConfigOptions, NodeStateTable<AnalyzerConfigOptionsProvider>.WithSingleItem(_state.OptionsProvider, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.ParseOptions, NodeStateTable<ParseOptions>.WithSingleItem(_state.ParseOptions, EntryState.Added));
+
+            // build the updated node for the syntax trees
+            PooledHashSet<SyntaxTree> syntaxTrees = PooledHashSet<SyntaxTree>.GetInstance();
+            foreach (var tree in compilation.SyntaxTrees)
             {
-                foreach (var tree in compilation.SyntaxTrees)
+                syntaxTrees.Add(tree);
+            }
+
+            var previousTrees = state.StateTable.GetStateTable<SyntaxTree>(SharedInputNodes.SyntaxTrees);
+            var newTrees = new NodeStateTable<SyntaxTree>.Builder();
+            foreach ((var tree, var treeState) in previousTrees)
+            {
+                newTrees.AddEntriesFromPreviousTable(previousTrees, syntaxTrees.Contains(tree) ? EntryState.Cached : EntryState.Removed);
+                syntaxTrees.Remove(tree);
+            }
+            foreach (var tree in syntaxTrees)
+            {
+                newTrees.AddEntries(ImmutableArray.Create(tree), EntryState.Added);
+            }
+            var syntaxTreeTable = newTrees.ToImmutableAndFree();
```

Comment:
>var syntaxTreeTable = newTrees.ToImmutableAndFree(); [](http://example.com/codeflow?start=12&length=52)

Could this whole calculation of `syntaxTreeTable` be a method called on the previous state table? (It's hard to follow here and it looks like there's an implied index used in `AddEntriesFromPreviousTable()` above.)

Perhaps a method such as:
```csharp
var syntaxTreeTable = previousTrees.GetUpdatedTable(compilation.SyntaxTrees);
```